### DISCUSSION TITLE
testing on rpi 4B+

### DISCUSSION THREAD

**Discussion by inglis-dl** [rickbassham/livestack#5] [Show and tell]
I ran a test of livestack running on a pi 4B+ 8Gb with SM OS 1.5.5.
The images are from some initial tests of a new ASI533MC Pro with shots of the Perseus double cluster
taken 2020-12-29 before and after a meridian flip.  10 darks taken with the same gain, 
exposure, temperature and offset were placed in the livestack input folder.  On startup, livestack created a master dark and I then scp'd 17 (curated previously in ASTAP) lights.  For each light, lightstack subtracts the master dark, debayers, aligns, stacks and stretches the resulting stacked image.  I ran a Siril script on the same set of dark and light data files for comparison, applied basic background neutralization and photometric color calibration and saved the auto-stretched result.

_livestack:_
![livestack_process](https://user-images.githubusercontent.com/722661/103380102-dcddee00-4ab5-11eb-8594-a9aca25efc18.png)

_siril:_
![siril_process_matchscale](https://user-images.githubusercontent.com/722661/103380275-5d9cea00-4ab6-11eb-8cfb-b70cc977cdff.png)

**Comment by rickbassham** [comment]
Not bad. Nice to see another setup work. livestack might be going a little heavy handed on the color saturation. I might dial that down a bit.

How was the performance on the RPi4 in livestack vs siril?

**Comment by rickbassham** [comment]
Also, thanks for the bug reports! Glad to see it working with FITS files generated by a different camera than what I own.

**Comment by inglis-dl** [comment]
siril is running on a i7-7700HQ 8 core / 16 Gb / Quadro M1200 graphics
laptop, livestack is on the pi 4+ 8 Gb with a 128Gb SSD drive.
I haven't done a timing but siril is orders of magnitude faster, and that
is fine, since livestack is processing lights as they
come off the camera, in this case it would have been every 30 seconds.
Interestingly, I had taken 5 shots with Ekos and named
the object 'DOUBLE_CLUSTER' and the other 12 shots named 'NGC_884' and
livestack "knew" based on fits header tag to
stack 2 separate sets of images.  I cheated it by chaning the fits header
object tag in all files to just 'NGC_884' and had livestack
do all 17 as one set.

**Reply by rickbassham** [reply]
Yeah, livestack looks at the `OBJECT` FITS header. So you could come back to a target, or do multiple targets in a night and it wouldn't try to stack over them.

**Comment by inglis-dl** [comment]
and thank-you for creating this software!!! I couldnt get it to come up in localhost:8080 in chrome running on the pi, and I couldnt get ekos-live to run within Ekos on the pi: it wouldn't accept any credentials I entered.  Here is what I hope to do with livestack within my workflow:
- run Kstars / Ekos on the pi (SM OS v1.5.5) with all gear attached to the pi / Pegasus PBA
- from inside the house I run RealVNC on my laptop over wifi to control Ekos: pi is wifi connected to an extender on the same lan (ie no stellarmate hotspot)
- Ekos stores images on the pi so I dont run into file transfer issues / lag over lan
- run livestack on the pi
- view the results on the laptop ...
How do I setup the last part?  Do I install/run  Ekos-Live from your docker container ?

**Reply by rickbassham** [reply]
I'm not sure ekos-web will work on stellarmate, since they have their own thing going. I built ekos-web to be a standalone version of their local web manager, since I run ubuntu and don't use stellarmate.

In the docker-compose file for livestack, try changing the line

```
      - "8080:80"
```
to something else, like (just don't change the `:80` part)
```
      - "8081:80"
```

Then `docker-compose up` again and see if you can view it at http://localhost:8081. Something else might be using port 8080 already.

**Reply by inglis-dl** [reply]
interesting ... same ERR_CONNECTION_REFUSED

**Reply by rickbassham** [reply]
Can you post your `docker-compose.yml` file?

**Reply by inglis-dl** [reply]
attached 
[docker-compose.zip](https://github.com/rickbassham/livestack/files/5755787/docker-compose.zip)

**Comment by rickbassham** [comment]
Alright, try this:

Make a copy of your docker-compose.yml file in case there are conflicts.

```bash
# this will undo any changes you've made to your docker-compose.yml file
git checkout docker-compose.yml
git pull
docker-compose pull nginx
# now update your livestack volumes to be what they need to be in docker-compose.yml
docker-compose up
```

**Comment by rickbassham** [comment]
Also, the terminal output should show some lines like `done processing file ... in XXXms`. I'd be interested in seeing those to gauge performance on the raspberry pi.

**Comment by inglis-dl** [comment]
timings on dark processing ... lights to come later

stellarmate@stellarmate:~/Documents/livestack $ docker-compose up   
Recreating livestack_livestack_1 ... done
Recreating livestack_nginx_1     ... done
Attaching to livestack_livestack_1, livestack_nginx_1
nginx_1      | /docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
nginx_1      | /docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
nginx_1      | /docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
nginx_1      | 10-listen-on-ipv6-by-default.sh: info: can not modify /etc/nginx/conf.d/default.conf (read-only file system?)
nginx_1      | /docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
nginx_1      | /docker-entrypoint.sh: Configuration complete; ready for start up
livestack_1  | INFO:root:start processing file /livestack/input/Dark/Dark_Dark_001.fits
livestack_1  | INFO:root:no reference found for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:start saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 424.537ms
livestack_1  | INFO:root:done processing file /livestack/input/Dark/Dark_Dark_001.fits in 1470.209ms
livestack_1  | INFO:root:9 items remaining
livestack_1  | INFO:root:start processing file /livestack/input/Dark/Dark_Dark_002.fits
livestack_1  | INFO:root:start stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 2215.465ms
livestack_1  | INFO:root:start saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 424.973ms
livestack_1  | INFO:root:done processing file /livestack/input/Dark/Dark_Dark_002.fits in 3836.028ms
livestack_1  | INFO:root:8 items remaining
livestack_1  | INFO:root:start processing file /livestack/input/Dark/Dark_Dark_003.fits
livestack_1  | INFO:root:start stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 2190.752ms
livestack_1  | INFO:root:start saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 426.515ms
livestack_1  | INFO:root:done processing file /livestack/input/Dark/Dark_Dark_003.fits in 3810.137ms
livestack_1  | INFO:root:7 items remaining
livestack_1  | INFO:root:start processing file /livestack/input/Dark/Dark_Dark_004.fits
livestack_1  | INFO:root:start stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 2198.277ms
livestack_1  | INFO:root:start saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 442.302ms
livestack_1  | INFO:root:done processing file /livestack/input/Dark/Dark_Dark_004.fits in 3823.148ms
livestack_1  | INFO:root:6 items remaining
livestack_1  | INFO:root:start processing file /livestack/input/Dark/Dark_Dark_005.fits
livestack_1  | INFO:root:start stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 2241.540ms
livestack_1  | INFO:root:start saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 425.349ms
livestack_1  | INFO:root:done processing file /livestack/input/Dark/Dark_Dark_005.fits in 3856.002ms
livestack_1  | INFO:root:5 items remaining
livestack_1  | INFO:root:start processing file /livestack/input/Dark/Dark_Dark_006.fits
livestack_1  | INFO:root:start stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 2201.810ms
livestack_1  | INFO:root:start saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 434.835ms
livestack_1  | INFO:root:done processing file /livestack/input/Dark/Dark_Dark_006.fits in 3824.118ms
livestack_1  | INFO:root:4 items remaining
livestack_1  | INFO:root:start processing file /livestack/input/Dark/Dark_Dark_007.fits
livestack_1  | INFO:root:start stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 2114.989ms
livestack_1  | INFO:root:start saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 432.327ms
livestack_1  | INFO:root:done processing file /livestack/input/Dark/Dark_Dark_007.fits in 3728.510ms
livestack_1  | INFO:root:3 items remaining
livestack_1  | INFO:root:start processing file /livestack/input/Dark/Dark_Dark_008.fits
livestack_1  | INFO:root:start stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 2223.171ms
livestack_1  | INFO:root:start saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 445.441ms
livestack_1  | INFO:root:done processing file /livestack/input/Dark/Dark_Dark_008.fits in 3842.729ms
livestack_1  | INFO:root:2 items remaining
livestack_1  | INFO:root:start processing file /livestack/input/Dark/Dark_Dark_009.fits
livestack_1  | INFO:root:start stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 2242.831ms
livestack_1  | INFO:root:start saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 433.575ms
livestack_1  | INFO:root:done processing file /livestack/input/Dark/Dark_Dark_009.fits in 3859.511ms
livestack_1  | INFO:root:1 items remaining
livestack_1  | INFO:root:start processing file /livestack/input/Dark/Dark_Dark_010.fits
livestack_1  | INFO:root:start stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done stacking image for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 2557.048ms
livestack_1  | INFO:root:start saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done saving stacked fits for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 503.467ms
livestack_1  | INFO:root:done processing file /livestack/input/Dark/Dark_Dark_010.fits in 4326.920ms
livestack_1  | INFO:root:0 items remaining

**Reply by rickbassham** [reply]
Nice. 3-4 seconds per image isn't bad. The lights will likely be longer due to debayering and aligning.

**Comment by inglis-dl** [comment]
here is from process of the first light ...

nginx_1      | 172.18.0.1 - - [30/Dec/2020:22:07:37 +0000] "GET / HTTP/1.1" 200 372 "-" "Mozilla/5.0 (X11; Linux armv7l) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.141 Safari/537.36" "-"
nginx_1      | 2020/12/30 22:07:37 [error] 21#21: *1 open() "/opt/www/files/favicon.ico" failed (2: No such file or directory), client: 172.18.0.1, server: localhost, request: "GET /favicon.ico HTTP/1.1", host: "localhost:8080", referrer: "http://localhost:8080/"
nginx_1      | 172.18.0.1 - - [30/Dec/2020:22:07:37 +0000] "GET /favicon.ico HTTP/1.1" 404 555 "http://localhost:8080/" "Mozilla/5.0 (X11; Linux armv7l) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.141 Safari/537.36" "-"
nginx_1      | 172.18.0.1 - - [30/Dec/2020:22:07:41 +0000] "GET /output/ HTTP/1.1" 200 160 "http://localhost:8080/" "Mozilla/5.0 (X11; Linux armv7l) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.141 Safari/537.36" "-"
nginx_1      | 172.18.0.1 - - [30/Dec/2020:22:07:43 +0000] "GET / HTTP/1.1" 200 372 "http://localhost:8080/output/" "Mozilla/5.0 (X11; Linux armv7l) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.141 Safari/537.36" "-"
nginx_1      | 172.18.0.1 - - [30/Dec/2020:22:07:45 +0000] "GET /storage/ HTTP/1.1" 200 437 "http://localhost:8080/" "Mozilla/5.0 (X11; Linux armv7l) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.141 Safari/537.36" "-"
nginx_1      | 172.18.0.1 - - [30/Dec/2020:22:07:49 +0000] "GET / HTTP/1.1" 200 372 "http://localhost:8080/storage/" "Mozilla/5.0 (X11; Linux armv7l) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.141 Safari/537.36" "-"
nginx_1      | 172.18.0.1 - - [30/Dec/2020:22:07:49 +0000] "GET /storage/ZWO%20CCD%20ASI533MC%20Pro_DARK_30.0_120.0_-15.fits HTTP/1.1" 200 36195840 "http://localhost:8080/storage/" "Mozilla/5.0 (X11; Linux armv7l) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.141 Safari/537.36" "-"
livestack_1  | INFO:root:file created: /livestack/input/Light/DOUBLE_CLUSTER_TEST_001.fits
livestack_1  | INFO:root:file created: /livestack/input/Light/DOUBLE_CLUSTER_TEST_002.fits
livestack_1  | INFO:root:start processing file /livestack/input/Light/DOUBLE_CLUSTER_TEST_001.fits
livestack_1  | INFO:root:start subtracting dark for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15
livestack_1  | INFO:root:done subtracting dark for ZWO CCD ASI533MC Pro_DARK_30.0_120.0_-15 in 338.883ms
livestack_1  | INFO:root:no flat found for ZWO CCD ASI533MC Pro_FLAT_NONE_120.0_-15
livestack_1  | INFO:root:start debayering image for ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15 with pattern RGGB
livestack_1  | INFO:root:done debayering image for ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15 with pattern RGGB in 7769.960ms
livestack_1  | INFO:root:no reference found for ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15
livestack_1  | INFO:root:no reference found for ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15
livestack_1  | INFO:root:start saving stacked fits for ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15
livestack_1  | INFO:root:file created: /livestack/input/Light/DOUBLE_CLUSTER_TEST_003.fits
livestack_1  | INFO:root:done saving stacked fits for ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15 in 3031.979ms
livestack_1  | INFO:root:start stretch
livestack_1  | INFO:root:done stretch in 3212.994ms
livestack_1  | INFO:root:start stretch
nginx_1      | 172.18.0.1 - - [30/Dec/2020:22:09:08 +0000] "GET / HTTP/1.1" 200 372 "http://localhost:8080/storage/" "Mozilla/5.0 (X11; Linux armv7l) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.141 Safari/537.36" "-"
livestack_1  | INFO:root:file created: /livestack/input/Light/DOUBLE_CLUSTER_TEST_004.fits
livestack_1  | INFO:root:done stretch in 3314.009ms
livestack_1  | INFO:root:start stretch
nginx_1      | 172.18.0.1 - - [30/Dec/2020:22:09:13 +0000] "GET /output/ HTTP/1.1" 200 160 "http://localhost:8080/" "Mozilla/5.0 (X11; Linux armv7l) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.141 Safari/537.36" "-"
nginx_1      | 172.18.0.1 - - [30/Dec/2020:22:09:13 +0000] "GET /output/ HTTP/1.1" 200 160 "http://localhost:8080/" "Mozilla/5.0 (X11; Linux armv7l) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.141 Safari/537.36" "-"
livestack_1  | INFO:root:done stretch in 3173.362ms
livestack_1  | INFO:root:start saving ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15.png
livestack_1  | INFO:root:done saving ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15.png in 340.742ms
livestack_1  | INFO:root:done processing file /livestack/input/Light/DOUBLE_CLUSTER_TEST_001.fits in 28715.668ms

**Comment by inglis-dl** [comment]
so now on my laptop I can remotely view the stacked png file in firefox but there are no UI elements, just a simple 
directory tree "Index of /" and I can navigate to and open the file in the browser page

**Reply by rickbassham** [reply]
Yeah, that's all there is for now. I might do up a simple auto-refreshing ui at some point like ekos-web, but at least you can see it. You'll need to refresh manually though. :(

**Comment by rickbassham** [comment]
Thanks again for all the testing on this. It's good to see it working on another system. I've added you as a contributor to the project as recognition.

**Reply by inglis-dl** [reply]
cheers! 

yep, nginx is registering my refresh hits:
nginx_1      | 192.168.0.109 - - [30/Dec/2020:22:23:43 +0000] "GET /output/ZWO%20CCD%20ASI533MC%20Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15.png HTTP/1.1" 304 0 "http://192.168.0.192:8080/output/" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:84.0) Gecko/20100101 Firefox/84.0" "-"

this is the tail end of the stack process:

livestack_1  | INFO:root:done aligning image for ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15 in 21831.393ms
livestack_1  | INFO:root:start stacking image for ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15
livestack_1  | INFO:root:done stacking image for ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15 in 1252.272ms
livestack_1  | INFO:root:start saving stacked fits for ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15
livestack_1  | INFO:root:done saving stacked fits for ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15 in 1870.617ms
livestack_1  | INFO:root:start stretch
livestack_1  | INFO:root:done stretch in 2784.830ms
livestack_1  | INFO:root:start stretch
livestack_1  | INFO:root:done stretch in 2886.747ms
livestack_1  | INFO:root:start stretch
livestack_1  | INFO:root:done stretch in 2837.646ms
livestack_1  | INFO:root:start saving ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15.png
livestack_1  | INFO:root:done saving ZWO CCD ASI533MC Pro_LIGHT_NGC_884_NONE_30.0_120.0_-15.png in 334.608ms
livestack_1  | INFO:root:done processing file /livestack/input/Light/DOUBLE_CLUSTER_TEST_017.fits in 50234.273ms
livestack_1  | INFO:root:0 items remaining

**Reply by rickbassham** [reply]
Oof. Almost a full minute to do a file. I'll see if I can get that down. Aligning is taking over 20 seconds, and debayering is around 8. And 3 seconds to stretch each channel... Got a ways to go before it is fast.
